### PR TITLE
Completed python data series.

### PR DESCRIPTION
**Pull request by kevinmbeck**
Screen Shot 2022-04-28 at 12 17 01 AM
Screen Shot 2022-04-28 at 1 15 50 AM
Screen Shot 2022-04-28 at 10 50 44 PM
Screen Shot 2022-04-30 at 11 27 40 AM

### CODE REVIEW

File: ds-prep-python-data-series/list_tuples.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    (2010, 11),
+    (2004, 6),
+    (2020, 4)
+]
+
+# Write your code below this line:
+def last(list):
+    d = dict(zip(columns, list[-1]))
+    return(d)
+print(last(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:38]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+def five(list):
+    return(list[:5])
+print(five(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:41]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+def included(year, list):
+    if [x for x in list if x[0] == year]:
+        return(True)
+    else:
+        return(False)
+print(included(2000, unemployment_rates))
+print(included(2010, unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:48]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+def recent(list):
+    list.sort()
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py]: `list.sort()` modifies the order in `list`. Please attempt to complete these without modifying the original data set.

File: ds-prep-python-data-series/list_tuples.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    l = list[-1]
+    return(l[1])
+print(recent(unemployment_rates))
+def years(list):
+    s= set()
+    for x in list:
+        s.add(x[0])
+    return(s)
+print(years(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:59]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+def rates(list):
+    list.sort
+    l = []
+    for x in list:
+        l.append(x[1])
+    return(l)
+print(rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py]: Please avoid `.sort`ing the original data set.

File: ds-prep-python-data-series/list_tuples.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+def largest(list):
+    list.sort(key=lambda x: x[1])
+    l = list[-1]
+    return(l[1])
+print(largest(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py]: Please avoid `.sort`ing the original data set.

File: ds-prep-python-data-series/list_tuples.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+def employment_rates(l):
+    l1 = []
+    l2 = []
+    for x in l:
+        l1.append(x[0])
+        l2.append(100-x[1])
+    l3 = list(zip(l1, l2))
+    return(l3)
+print(employment_rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:80]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+def greater(list):
+    l = []
+    for x in list:
+        if x[1]>= 7:
+            l.append(x)
+    return(l)
+print(greater(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:87]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+def number(list):
+    c = dict()
+    for x in list:
+        if (x[1] in c):
+            c[x[1]] += 1
+        else:
+            c[x[1]] = 1
+    return(c)
+print(number(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:96]: ✔

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]
Feedback for `list_tuples.py`, so you can get started on it. Feedback on other files coming up next.

File: ds-prep-python-data-series/list_dicts.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    {"unemployment_rate": 4, "year": 2020},
+    {"unemployment_rate": 10, "year": 2011},
+    {"unemployment_rate": 8, "year": 2009},
+    {"unemployment_rate": 5, "year": 2008}
+]
+
+# Write your code below this line:
+def last(list):
+    return(list[-1])
+print(last(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:37]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+def five(list):
+    c = [tuple(x.values()) for x in list]
+    return(c[:5])
+print(five(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:41]: ✔

**Comment by kevinmbeck** [comment]
Screen Shot 2022-04-30 at 12 06 23 PM
Fixed sort problems in list_tuples. I also have sort problems in other files so I'll fix those too

**Comment by kevinmbeck** [comment]
I realized I might have another issues of modifying original data sets. I changed the original data sets in everything except list_tuples to get the employment rate. Which is why I have the employment rate function at the very end. But I tried copying the list to a new list and then changing the unemployment rate values in just the copy but it still changed the original data set and I'm not sure why

File: ds-prep-python-data-series/list_dicts.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+def included(year, list):
+    for x in list:
+        if any(x["year"] == year for x in list):
+            return(True)
+        else:
+            return(False)
+print(included(2000, unemployment_rates))
+print(included(2010, unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:49]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+def recent(list):
+    l1 = sorted(list, key = lambda item: item.get("year"))
+    l = l1[-1]
+    return(l.get("unemployment_rate"))
+print(recent(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:54]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+def years(list):
+    s= set()
+    for x in list:
+        s.add(x["year"])
+    return(s)
+print(years(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:60]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+def rates(list):
+    l1= sorted(list, key = lambda item: item.get("year"))
+    l = []
+    for x in l1:
+        l.append(x["unemployment_rate"])
+    return(l)
+print(rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:67]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+def largest(list):
+    l1 =sorted(list, key=lambda x: x.get("unemployment_rate"))
+    l = l1[-1]
+    return(l.get("unemployment_rate"))
+print(largest(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:72]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+def greater(list):
+    l = []
+    for x in list:
+        if x["unemployment_rate"]>= 7:
+            l.append(x)
+    return(l)
+print(greater(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:79]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+def number(list):
+    c = dict()
+    for x in list:
+        if (x["unemployment_rate"] in c):
+            c[x["unemployment_rate"]] += 1
+        else:
+            c[x["unemployment_rate"]] = 1
+    return(c)
+print(number(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:88]: ✔

File: ds-prep-python-data-series/list_dicts.py [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+def employment_rates(list):
+    copied = list.copy()
+    for x in copied:
+        x.update({"employment_rate": 100-x.get("unemployment_rate")})
+        del x["unemployment_rate"]
+    return(copied)
+print(employment_rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:95]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+        return(True)
+    else:
+        return(False)
+print(included(2000, unemployment_rates))
+print(included(2010, unemployment_rates))
+def recent(list):
+    l = sorted(list)
+    l1 = l[-1]
+    return(l1[1])
+print(recent(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:53]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+        s.add(x[0])
+    return(s)
+print(years(unemployment_rates))
+def rates(list):
+    l1 = sorted(list)
+    l = []
+    for x in l1:
+        l.append(x[1])
+    return(l)
+print(rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:66]: ✔

File: ds-prep-python-data-series/list_tuples.py [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+def largest(list):
+    l1 = sorted(list, key=lambda x: x[1])
+    l = l1[-1]
+    return(l[1])
+print(largest(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:71]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+    2017: 5,
+    2003: 6
+}
+
+# Write your code below this line:
+def last(di):
+    c = list(di.items())
+    d = dict(zip(columns, c[-1]))
+    return(d)
+print(last(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:39]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+def five(dict):
+    c = list(dict.items())
+    return(c[:5])
+print(five(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:43]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+def included(year, dict):
+    if year in dict:
+        return(True)
+    else:
+            return(False)
+print(included(2000, unemployment_rates))
+print(included(2010, unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:50]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+def recent(dict):
+    c = list(dict.items())
+    c.sort()
+    l = c[-1]
+    return(l[1])
+print(recent(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:56]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+def years(dict):
+    s= set()
+    for x in dict:
+        s.add(x)
+    return(s)
+print(years(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:62]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+def rates(dict):
+    c = list(dict.items())
+    c.sort()
+    l = []
+    for x in c:
+        l.append(x[1])
+    return(l)
+print(rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:70]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+def largest(dict):
+    c = list(dict.items())
+    c.sort(key=lambda x: x[1])
+    l = c[-1]
+    return(l[1])
+print(largest(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:76]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+def greater(d):
+    d1 = {}
+    for x, y in d.items():
+        if y >= 7:
+            d1[x] = y
+    return(d1)
+print(greater(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:83]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+def number(d):
+    c = dict()
+    for x, y in d.items():
+        if (y in c):
+            c[y] += 1
+        else:
+            c[y] = 1
+    return(c)
+print(number(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:92]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+def employment_rate(dict):
+    for x, y in dict.items():
+        dict.update({x: 100-y})
+    return(dict)
+print(employment_rate(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py]: Please don't modify the original data set.

File: ds-prep-python-data-series/dict_columns.py [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+        6
+    ]
+}
+
+# Write your code below this line:
+def last(di):
+    c = list(zip(di['year'], di['unemployment_rate']))
+    d = dict(zip(columns, c[-1]))
+    return(d)
+print(last(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:32]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+def five(dict):
+    c = list(zip(dict['year'], dict['unemployment_rate']))
+    return(c[:5])
+print(five(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:36]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+def included(year, dict):
+    if year in dict['year']:
+        return(True)
+    else:
+        return(False)
+print(included(2000, unemployment_rates))
+print(included(2010, unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:43]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 37]
```diff
@@ -0,0 +1,100 @@
...
+def recent(dict):
+    c = list(zip(dict['year'], dict['unemployment_rate']))
+    c.sort()
+    l = c[-1]
+    return(l[1])
+print(recent(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:49]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+def years(dict):
+    s= set()
+    for x in dict['year']:
+        s.add(x)
+    return(s)
+print(years(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:55]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 39]
```diff
@@ -0,0 +1,100 @@
...
+def rates(dict):
+    c = list(zip(dict['year'], dict['unemployment_rate']))
+    c.sort()
+    l = []
+    for x in c:
+        l.append(x[1])
+    return(l)
+print(rates(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:63]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 40]
```diff
@@ -0,0 +1,100 @@
...
+def largest(dict):
+    c = list(zip(dict['year'], dict['unemployment_rate']))
+    c.sort(key=lambda x: x[1])
+    l = c[-1]
+    return(l[1])
+print(largest(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:69]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+    l1 = []
+    l2 = []
+    for x in l:
+        l1.append(x[0])
+        l2.append(x[1])
+    d = {}
+    d['year'] = l1
+    d['unemployment_rate'] = l2
+    return(d)
+print(greater(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:85]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 42]
```diff
@@ -0,0 +1,100 @@
...
+def number(d):
+    c = dict()
+    for x in d['unemployment_rate']:
+        if (x in c):
+            c[x] += 1
+        else:
+            c[x] = 1
+    return(c)
+print(number(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:94]: ✔

File: ds-prep-python-data-series/dict_columns.py [hunk 43]
```diff
@@ -0,0 +1,100 @@
...
+def employment_rate(dict):
+    dict['employment_rate'] = dict.pop('unemployment_rate')
+    for x in dict['employment_rate']:
+        dict['employment_rate'] = [100-x for x in dict['employment_rate']]
+    return(dict)
+print(employment_rate(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py]: Please don't modify the original data set.

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]
Almost there, just a couple changes

File: ds-prep-python-data-series/dict_columns.py [hunk 44]
```diff
@@ -0,0 +1,101 @@
...
+            c[x] = 1
+    return(c)
+print(number(unemployment_rates))
+def employment_rate(dict):
+    copied = dict.copy()
+    copied['employment_rate'] = copied.pop('unemployment_rate')
+    for x in copied['employment_rate']:
+        copied['employment_rate'] = [100-x for x in copied['employment_rate']]
+    return(copied)
+print(employment_rate(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_columns.py:101]: ✔

File: ds-prep-python-data-series/dict_labels.py [hunk 45]
```diff
@@ -0,0 +1,98 @@
...
+        else:
+            c[y] = 1
+    return(c)
+print(number(unemployment_rates))
+def employment_rate(dict):
+    copied = dict.copy()
+    for x, y in copied.items():
+        copied.update({x: 100-y})
+    return(copied)
+print(employment_rate(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/dict_labels.py:98]: ✔

**Review by TimelikeClosure — APPROVED** [review]

**Pull request merged**